KHR_texture_procedurals Draft Proposal

kwokcb opened the pull request:
## Draft Proposal for `KHR_texture_procedurals`

- [x] Specification proposal ready for initial feedback.
- [ ] Draft `JSON` Schema is still progress.
- [ ] Example assets will be forthcoming in a [glTF Sample Assets](https://github.com/KhronosGroup/glTF-Sample-Assets) branch.

kwokcb committed:
Branch for KHR_texture_procedurals

kwokcb committed:
Merge branch 'main' into KHR_texture_procedurals

kwokcb committed:
Transfer over working document (#2376)

* Start document consolidation.

* Transferred doc.

kwokcb committed:
Cleanup spec document.

kwokcb committed:
KHR_texture_procedurals:
- Update checkerboard example with USD equiivalent and mermaid graph.
- Add USD/ / MaterialX / glTF mapping

kwokcb committed:
Finish transfer of google docs information (minus verbose examples) to spec README.

kwokcb committed:
Merge branch 'main' into KHR_texture_procedurals

ashwinbhat commented on extensions/2.0/Khronos/KHR_texture_procedurals/README.md:
```diff
@@ -0,0 +1,1397 @@
+
+1. **Interoperability**: Adhere to an industry a standard node schema (MaterialX) with a runtime friendly JSON representation. 
+
+    More specifically __only__ nodes defined in the MaterialX specification are supported. This ensures that the procedural graphs can be easily converted to MaterialX and USDShade graphs.
```
MaterialX "1.39" specification

ashwinbhat replied:
Second extension will be written against gltf 2.0 and KHR_procedural

ashwinbhat commented on extensions/2.0/Khronos/KHR_texture_procedurals/README.md:
```diff
@@ -0,0 +1,1397 @@
+
+    For the first version of this extension nodes which are used to define shading models are not allowed. Please refer to the [resources](#resources) section for links to supported  MaterialX node definitions.
+
+2. **Fidelity**: Provide the ability to generate complex patterns, noise, or other effects that currently must be "baked" into texture maps. Provide the ability to provide mappings from one shading model to another without baking in a consistent manner via procedural graphs. Provide the ability to support NPR effects for unlit materials. Reduces runtime memory usage by generating patterns / shadi ...
```
"Provide the ability to support NPR effects for unlit materials" - is this required?

ashwinbhat commented on extensions/2.0/Khronos/KHR_texture_procedurals/README.md:
````diff
@@ -0,0 +1,1397 @@
+``` 
+with the extension object defined in the `extensions array:
+
+```json
+{
+    "extensions": {
+        "KHR_texture_procedurals": {
+            "mimetype": "application/mtlx+json;version=<MaterialX_version>",
+            "procedurals": []
+        }
+    }
+}
````
Following up from last call, proposed change is to simply mention
```
"extensions": {
        "KHR_texture_procedurals"
    }
```
Which indicates that gltf texture procedural "structure" is used. 
When gltf texture procedural uses MaterialX 1.39 nodes the extension will be 
```
"extensions": 
                "KHR_texture_procedurals_MaterialX_1.39": {
}
```
Example usage:
```
"extensionsRequired" : [
    "KHR_texture_procedurals"
]
"extensionsUsed" : [
    "KHR_texture_procedurals_MaterialX_1.39"
]

```

ashwinbhat replied:
use case #1 gltf using Mtlx 1.39
extensionsRequired" : [  
    "KHR_texture_procedurals"  
    "KHR_texture_procedurals_MaterialX_1.39"  
]  
use case #2 gltf using Mtlx 2.0
extensionsRequired" : [  
    "KHR_texture_procedurals"  
    "KHR_texture_procedurals_MaterialX_2.00"
]  
use case #3
extensionsRequired" : [  
    "KHR_texture_procedurals"  
    "NV_MDL_texture_procedurals"  
]  
(mixing of procedurals is not allowed, only one node library can be used at a time)
extensionsRequired" : [  
    "KHR_texture_procedurals"  
    "KHR_texture_procedurals_MaterialX_1.39"  
    "KHR_texture_procedurals_MaterialX_2.00"      
]

kwokcb replied:
What does it mean to include `NV_MDL_texture_procedurals` is there some co-dependency implied here or just an example ?

kwokcb replied:
What does the `extenions` portion look like. Is it still this ?
```json
{
    "extensions": {
        "KHR_texture_procedurals": {
            "procedurals": []
        }
    }
}
```
Or do you need something like this ?
```json
{
    "extensions": {
        "KHR_texture_procedurals_MaterialX_1.39": {
            "procedurals": []
        }
    }
}
```

ashwinbhat commented on extensions/2.0/Khronos/KHR_texture_procedurals/README.md:
````diff
@@ -0,0 +1,1397 @@
+        }
+    }
+}
+```
+It is assumed that a mimetype is always required. As part of the mimetype a version `<MaterialX_version>` is specified. This is the version of the MaterialX library specification used when writing to glTF. The version is specified as a string in the form of `<major version>.<minor version>`. For example, if the MaterialX library version is 1.39, the mimetype would be `application/mtlx+json;versio ...
+
+The correspond version is specified in a MaterialX XML document as follows:
+
+```xml
+<materialx version="1.39">
+</materialx>
+```
+
+The version in the mimetype __is not__ the extension version. If in the future the schema needs to be modified (perhaps due to changes in MaterialX) then a new extension version would be required.
+
+The `procedurals` array specifies the procedural graphs that are used in the glTF asset. 
+
+### Versioning
+
+Elements from different versions of MaterialX is disallowed. External tooling must handle this (e.g. by performing an "upgrade" operation).
+
+<img src="./figures//version_upgrade.svg" width=100%>
+
````
This section can be removed since use of KHR_texture_procedurals + KHR_texture_procedurals_MaterialX_1.39 will imply MaterialX library used.

kwokcb committed:
Update versioned extension declaration syntax.

kwokcb commented on extensions/2.0/Khronos/KHR_texture_procedurals/README.md:
````diff
@@ -0,0 +1,1400 @@
+```json
+{
+    "extensions": {
+        "KHR_texture_procedurals_MaterialX_1.39": {
````
Is this the correct usage ? i.e. the extension is hard-coded to a version ?

ashwinbhat committed:
Update to use abbreviated name for MaterialX extension

ashwinbhat committed:
Separate into two proposals
1. KHR_texture_procedural
2. EXT_texture_procedural_mx_1_39

ashwinbhat committed:
Fixup some links

kwokcb committed:
Merge branch 'main' into KHR_texture_procedurals

kwokcb committed:
Cleanup EXT_texture_procedurals_mx_1.39 for extension specification. Add in mapping section for texture_infos.

rsahlin commented on extensions/2.0/Khronos/KHR_texture_procedurals/README.md:
```diff
@@ -0,0 +1,1351 @@
+
+### Motivation
+
+Textures represented as procedural graphs provides a way to extend the capabilities of glTF materials beyond what is possible with traditional texture maps. Key objectives includes:
```
Please reword to:
`Images represented as procedural operators or functions provides a way to extend the capabilities of glTF textures beyond what is possible with bitmap based images.`

rsahlin commented on extensions/2.0/Khronos/KHR_texture_procedurals/README.md:
```diff
@@ -0,0 +1,1351 @@
+
+    For the first version of this extension nodes which are used to define shading models are not allowed. Please refer to the [resources](#resources) section for links to supported  MaterialX node definitions.
+
+2. **Fidelity**: Provide the ability to generate complex patterns, noise, or other effects that currently must be "baked" into texture maps. Provide the ability to provide mappings from one shading model to another without baking in a consistent manner via procedural graphs. Provide the ability to map procedurals graphs to unlit materials. Reduces runtime memory usage by generating patterns / sha ...
```
I propose something along the lines of:

Provide the ability to generate procedural patterns, noise or other visual effects of sufficiently high fidelity to replace the use of pixel (bitmap) based images.
Reduce the transfer size by generating images, either at load- or run-time.

rsahlin commented on extensions/2.0/Khronos/KHR_texture_procedurals/README.md:
```diff
@@ -0,0 +1,1351 @@
+
+2. **Fidelity**: Provide the ability to generate complex patterns, noise, or other effects that currently must be "baked" into texture maps. Provide the ability to provide mappings from one shading model to another without baking in a consistent manner via procedural graphs. Provide the ability to map procedurals graphs to unlit materials. Reduces runtime memory usage by generating patterns / sha ...
+
+3. **Editability and Extensibility**: Extend runtime editability by exposing logic and interfaces for procedural graphs as well as providing a means to create new or extend existing node definitions.
```
This is not the goal of this extension - this extension should simply reference already existing operators/functions and provide a means to store them with a glTF asset.

rsahlin commented on extensions/2.0/Khronos/KHR_texture_procedurals/README.md:
```diff
@@ -0,0 +1,1351 @@
+
+### Material Binding
+
+To connect a graph `output` to a surface or displacement shader input the procedural extension can be declared within the a texture reference for a given material in the `materials` array.
```
I propose to simply connect the procedurals to an image in the `images`array - if this image is created at load time or dynamically created at runtime is up to the implementation.

rsahlin commented on extensions/2.0/Khronos/KHR_texture_procedurals/README.md:
```diff
@@ -0,0 +1,1351 @@
+</details>
+
+
+### Procedural Definitions
```
I propose a solution where the procedural image extension uses a set of already declared operators/functions.
These functions are what are available, if need for future additions that will need to be done in a separate extension.

The upside of having pre-declared operators is that its straightforward to do implementation testcases to verify if an implementation is compliant or not.
Ie, for each operator there are a number of testcases that provides input data and are used to validate the output.

The total amount of operators needed is not that large, probably around 20, maybe even less.

rsahlin replied:
For a  non complete comparison - have a look at what we specified in the Ikea procedural prototypes
This is not the complete declaration but it's enough to show the differences in design:

https://github.com/idlabs/procedural-image

rsahlin commented:
Food for thought!

Perhaps it's better to make this into a vendor extension (remove KHR prefix) and rename to `materialx_shader_graph` or something to that effect?

That way you would be able to keep the current design, if the extension receives traction it can be made into a KHR extension in the future.

What do you think?

kwokcb committed:
Update language for texture coordinate / transform bindings.

hybridherbst committed:
Fix typo in "multioutput" (#2432)

Thanks for the fix.

kwokcb committed:
Update on multi-channel texture_info and 3-channel texcoord bindings.

kwokcb committed:
Add clarifications for texute placement and sampling. All information comes from the nodes.
Inputs on nodes can be aniated. Further discussion on defails on existing extension required.
Explicit note N:M graph -> material bindings possible.

kwokcb committed:
Merge update. (#2440)

* KHR_materials_ior: Minor typo: plural instead of singular (#2413)

* Update anisotropy extension spec (#2409)

* Clarify and simplify formula for attenuation. (#2414)

* Add TRACE prefix (#2417)

* Update clearcoat extension spec (#2415)

* Update README. (#2427)

* Add KHR_node_hoverability (#2428)

* Add KHR_node_hoverability
* Resort extension list

* EXT_mesh_gpu_instancing: Suggest extensionRequired usage (#2405)

* Add button/link to glTF Discord (#2429)

As there is a lot more traffic to the Khronos glTF Discord than there is to Slack, a button pointing users to Discord will be helpful.

---------

Co-authored-by: Andreas Atteneder <andreas.atteneder@unity3d.com>
Co-authored-by: Alexey Knyazev <3479527+lexaknyazev@users.noreply.github.com>
Co-authored-by: Ed Mackey <elm19087@gmail.com>
Co-authored-by: Marco Hutter <javagl@javagl.de>
Co-authored-by: Khronos Group Web Services <webservices@khronosgroup.org>
Co-authored-by: Arseny Kapoulkine <arseny.kapoulkine@gmail.com>
Co-authored-by: James Riordon <outofcontrol@users.noreply.github.com>

kwokcb committed:
Fix schemas: remove mimetype and added in extensionsUsed.

kwokcb committed:
Fix vector to support matrix

Change max from 4 to 16 for data vector so
3x3 and 4x4 matrices can be supported.

kwokcb committed:
Update animation binding using animation pointer extension.

kwokcb committed:
Merge branch 'main' into KHR_texture_procedurals

DennisSmolek commented:
> Perhaps it's better to make this into a vendor extension (remove KHR prefix) and rename to `materialx_shader_graph` or something to that effect?

The only thing I think confusing is if this is a procedural texture for generating images to act as textures in existing materials (like your edits suggest) or is this for full procedural MATERIALS.

I will say, they way it is being advertised and promoted in the blogs and talks like Siggraph is as a procedural material, particularly as a channel for MaterialX materials.

One of the major goals listed for 2025 was to maintain good relations/interoperability with USD which uses MaterialX.

The way the extension is written now is different procedural systems could be used, be it other MatX versions or future systems like THREE TSL to generate the graph.

IMO the ship has sailed on this being texture only, and as it's a core goal I think the KHR name would stick. perhaps `KHR_procedural_materials` or whatever is more appropriate. 

---

PS if this is an old idea discussed elsewhere I apologize, it's hard finding the right channels to discuss these things.

DennisSmolek commented on extensions/2.0/Khronos/KHR_texture_procedurals/README.md:
```diff
@@ -0,0 +1,1462 @@
+    * `vector2`, `vector3`, `vector4` : 2, 3, 4 channel float vector
+    * `integer2`, `integer3`, `integer4` : 2, 3, 4 channel integer vector
+* matrix:
+    * `matrix3x3`, and `matrix4x4` : Row-major matrices of floats of size 3 or 4.
```
What nodes use strings as inputs/outputs?

DennisSmolek replied:
Yeah, at the moment I couldn't think of how nodes would use strings which now that I go through it there's quite a few, mostly filename and mode switches but also layers and other name specific stuff.
MaterialX itself uses a lot of name references which kinda breaks in JSON but I think most of them will work fine

kwokcb committed:
- Update input / outputs to be object sets vs arrays.
- Make single values be arrays.
- Update comparison table.
- Update and cleanup on example.

kwokcb committed:
Update diagrams to match latest spec changes.

kwokcb committed:
- Update colorspace, units, general meta-data support and definitions.
- For definitions add in disallowing defaults outputs for pass-through and implicit default geometry bindings.
- Add note about "targets" for definitions presuming these are disallowed.
- Add note about enumerations presuming these are discouraged or disallowed.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/KhronosGroup/glTF?pullRequest=2381) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/KhronosGroup/glTF?pullRequest=2381) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: kwokcb<br/>:white_check_mark: hybridherbst<br/>:x: ashwinbhat<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/KhronosGroup/glTF?pullRequest=2381) it.</sub>
